### PR TITLE
Efficiency improvements to the ADC module

### PR DESCRIPTION
**Pull request by Samragni-93**
This PR contributes efficiency improvements and new capabilities to the ADC module that supports calculation of IP/EA of closed- and open-shell molecules using three algorithms: in-core, out-of-core and density-fitting. This work is described in a recent ArXiV (https://arxiv.org/abs/2012.06714). The modifications to the code are as follows:

1. Improvement to the out-of-core algorithm for handling contractions of the OVVV and VVVV two-electron integrals
2. Improved approach for calculating the MP3 energy
3. More extensive use of efficient tensor intermediates throughout the code
4. A new analyze() function that performs the analysis of the ADC eigenvectors and spectroscopic factors
5. Improved interface that provides direct access to the IP/EA-ADC spectroscopic amplitudes
6. A new function that allows to calculate the IP/EA-ADC Dyson orbitals

### CODE REVIEW

**Comment by obackhouse** [comment]
Quick question about this PR, and regarding this line in your recent arXiv:

> As a result, EA/IP-RADC(3) and EA/IP-UADC(3) require a total of two and six O(o^2v^4) contractions, respectively, independent of the number of iterations in the Davidson algorithm.

Is there any reason not to store the result of `contract_ladder` between the first-order amplitudes and the (vv|vv) ERI block? It seems to me that, for example in the restricted code, the exact same contraction (at least with real integral symmetries) can be used for the contribution of this term to both the second-order amplitude and the third-order contribution to the 1p/1h space. Could it be stored as an intermediate i.e. `eris._t2_vvvv`, and freed once the 1p/1h space is built?

**Comment by ssh2** [comment]
> Quick question about this PR, and regarding this line in your recent arXiv:
> 
> > As a result, EA/IP-RADC(3) and EA/IP-UADC(3) require a total of two and six O(o^2v^4) contractions, respectively, independent of the number of iterations in the Davidson algorithm.
> 
> Is there any reason not to store the result of `contract_ladder` between the first-order amplitudes and the (vv|vv) ERI block? It seems to me that, for example in the restricted code, the exact same contraction (at least with real integral symmetries) can be used for the contribution of this term to both the second-order amplitude and the third-order contribution to the 1p/1h space. Could it be stored as an intermediate i.e. `eris._t2_vvvv`, and freed once the 1p/1h space is built?

Good point! This is probably an overlook on our side as a result of trying to keep the ground- and excited-state kernels relatively independent. I think you're right though and there is no reason why we cannot store the ladder intermediate after its calculated in the ground-state calculation and reuse it in the excited-state calculation. I'll let @Samragni-93 have a look at it and confirm, but it seems to me that it should be very easy to address as a part of this PR.

Thanks for the comment!

**Comment by Samragni-93** [comment]
> > Quick question about this PR, and regarding this line in your recent arXiv:
> > > As a result, EA/IP-RADC(3) and EA/IP-UADC(3) require a total of two and six O(o^2v^4) contractions, respectively, independent of the number of iterations in the Davidson algorithm.
> > 
> > 
> > Is there any reason not to store the result of `contract_ladder` between the first-order amplitudes and the (vv|vv) ERI block? It seems to me that, for example in the restricted code, the exact same contraction (at least with real integral symmetries) can be used for the contribution of this term to both the second-order amplitude and the third-order contribution to the 1p/1h space. Could it be stored as an intermediate i.e. `eris._t2_vvvv`, and freed once the 1p/1h space is built?
> 
> Good point! This is probably an overlook on our side as a result of trying to keep the ground- and excited-state kernels relatively independent. I think you're right though and there is no reason why we cannot store the ladder intermediate after its calculated in the ground-state calculation and reuse it in the excited-state calculation. I'll let @Samragni-93 have a look at it and confirm, but it seems to me that it should be very easy to address as a part of this PR.
> 
> Thanks for the comment!

Made changes and fixed this issue (commit : [d08393c](https://github.com/pyscf/pyscf/pull/795/commits/d08393c427a98672fcd5bd8f574996f18877fe39). The PR is ready for merge now.

File: pyscf/adc/dfadc.py [hunk 1]
```diff
@@ -15,6 +15,47 @@
...
+    ovvv = ovvv.reshape(-1, nvir, nvir, nvir)
+    return ovvv    
+
+
+def  get_ovvv_a_df(myadc, Lov, Lvv, p, chnk_size):
+
+    norb_1 = Lov.shape[1]
+    norb_2 = Lov.shape[2]
+    norb_3 = Lvv.shape[1]
+    naux = myadc.with_df.get_naoaux()
```
Comment (susilehtola) [inline pyscf/adc/dfadc.py:42]: You might want to use more descriptive variable names. It seems to me that `norb_1 = Nocc`, `norb_2 = Nvirt`, and `norb_3 = Nvirt`

File: pyscf/adc/__init__.py [hunk 2]
```diff
@@ -24,25 +24,28 @@
...
+    if isinstance(mf, scf.uhf.UHF):
         return UADC(mf, frozen, mo_coeff, mo_occ)
+    #elif isinstance(mf, scf.rohf.ROHF):
+    #    lib.logger.warn(mf, 'RADC method does not support ROHF reference. ROHF object '
+    #                    'is converted to UHF object and UADC method is called.')
+    #    mf = scf.addons.convert_to_uhf(mf)
+    #    return UADC(mf, frozen, mo_coeff, mo_occ)
+    # TODO add ROHF functionality
+    elif isinstance(mf, scf.rhf.RHF):
+        return RADC(mf, frozen, mo_coeff, mo_occ)
```
Comment (susilehtola) [inline pyscf/adc/__init__.py:44]: It seems to me these changes are spurious. You've merely changed the order of RHF and UHF and disabled ROHF altogether. I suggest resetting the file to the original version.

File: pyscf/adc/dfadc.py (hunk 1 above)
Comment (susilehtola) [reply pyscf/adc/dfadc.py:42]: You actually use better variable names above in `get_ovvv_df`

File: pyscf/adc/dfadc.py [hunk 3]
```diff
@@ -15,6 +15,47 @@
 
 import numpy as np
 
+def  get_ovvv_df(myadc, Lov, Lvv, p, chnk_size):
```
Comment (susilehtola) [inline pyscf/adc/dfadc.py:18]: Might want to add a docstring.

File: pyscf/adc/dfadc.py [hunk 4]
```diff
@@ -15,6 +15,47 @@
...
+    else :
+        Lov_temp = np.ascontiguousarray(Lov.transpose(1,2,0).reshape(-1,naux))
+
+    ovvv = np.dot(Lov_temp, Lvv)
+    ovvv = ovvv.reshape(-1, nvir, nvir, nvir)
+    return ovvv    
+
+
+def  get_ovvv_a_df(myadc, Lov, Lvv, p, chnk_size):
+
```
Comment (susilehtola) [inline pyscf/adc/dfadc.py:38]: Docstring missing here, too: how is this different from `get_ovvv_df`?

File: pyscf/adc/radc.py [hunk 5]
```diff
@@ -49,43 +50,47 @@ def kernel(adc, nroots=1, guess=None, eris=None, verbose=None):
...
-    if adc.verbose >= logger.INFO:
-        if nroots == 1:
-            logger.info(adc, '%s root %d    Energy (Eh) = %.10f    Energy (eV) = %.8f    Spec factors = %.8f    conv = %s',
-                         adc.method, 0, E, E*27.2114, spec_factors, conv)
-        else :
-            for n, en, pn, convn in zip(range(nroots), E, spec_factors, conv):
-                logger.info(adc, '%s root %d    Energy (Eh) = %.10f    Energy (eV) = %.8f    Spec factors = %.8f    conv = %s',
-                          adc.method, n, en, en*27.2114, pn, convn)
-        log.timer('ADC', *cput0)
+        logger.info(adc, "\n WARNING : Davidson iterations for " + str(nfalse) + " root(s) not converged\n")
```
Comment (susilehtola) [inline pyscf/adc/radc.py:67]:
```suggestion
        logger.warn(adc, "\n WARNING : Davidson iterations for " + str(nfalse) + " root(s) not converged\n")
```

File: pyscf/adc/radc.py [hunk 6]
```diff
@@ -94,47 +99,70 @@ def compute_amplitudes(myadc, eris):
...
-    # Compute first-order doubles t2 (tijab)
-
-    v2e_oovv = eris_ovov[:].transpose(0,2,1,3).copy()
-
     t2_1 = v2e_oovv/D2
+    if not isinstance(eris.oooo, np.ndarray):
+        t2_1 = radc_ao2mo.write_dataset(t2_1)
+        
+    del (v2e_oovv)
+    del (D2)
```
Comment (susilehtola) [inline pyscf/adc/radc.py:124]:
```suggestion
    del v2e_oovv
    del D2
```

File: pyscf/adc/radc.py [hunk 7]
```diff
@@ -94,47 +99,70 @@ def compute_amplitudes(myadc, eris):
...
-    eris_ovov = eris.ovov
     eris_oovv = eris.oovv
     eris_ovvo = eris.ovvo
-    eris_ovvv = eris.ovvv
 
-    e = myadc.mo_energy
+    # Compute first-order doubles t2 (tijab)
 
-    d_ij = e[:nocc][:,None] + e[:nocc]
+    v2e_oovv = eris_ovvo[:].transpose(0,3,1,2).copy()
```
Comment (susilehtola) [inline pyscf/adc/radc.py:107]: Dumb question: this is conversion from chemists' to physicists' notation?

Comment (susilehtola) [reply pyscf/adc/radc.py:107]: Does spliced `transpose` not already return a copy?

File: pyscf/adc/radc.py [hunk 8]
```diff
@@ -94,47 +99,70 @@ def compute_amplitudes(myadc, eris):
...
+    del (D2)
+
+    cput0 = log.timer_debug1("Completed t2_1 amplitude calculation", *cput0)
 
     # Compute second-order singles t1 (tij)
 
-    t2_1_a = t2_1 - t2_1.transpose(1,0,2,3).copy()
+    if isinstance(eris.ovvv, type(None)):
+        chnk_size = radc_ao2mo.calculate_chunk_size(myadc)
+    else :
```
Comment (susilehtola) [inline pyscf/adc/radc.py:132]:
```suggestion
    else:
```

You might also want to run the files through a formatter like YAPF to unify the code style.

File: pyscf/adc/radc.py [hunk 9]
```diff
@@ -391,19 +466,20 @@ def contract_ladder(myadc,t_amp,vvvv):
...
             k = vvvv_p.shape[0]
             vvvv_p = vvvv_p.reshape(-1,nvir*nvir)
-            t[a:a+k] = np.dot(vvvv_p,t_amp_t).reshape(-1,nvir,nocc*nocc)
+            t[a:a+k] = np.dot(vvvv_p,t_amp).reshape(-1,nvir,nocc*nocc)
             del (vvvv_p)
             a += k
     else :
         raise Exception("Unknown vvvv type") 
 
+    del (t_amp)
```
Comment (susilehtola) [inline pyscf/adc/radc.py:482]:
```suggestion
    del t_amp
```

File: pyscf/adc/radc.py [hunk 10]
```diff
@@ -391,19 +466,20 @@ def contract_ladder(myadc,t_amp,vvvv):
...
+             t[a:a+k] = np.dot(dataset,t_amp).reshape(-1,nvir,nocc*nocc)
              a += k
     elif getattr(myadc, 'with_df', None):
         for p in range(0,nvir,chnk_size):
             vvvv_p = dfadc.get_vvvv_df(myadc, vvvv, p, chnk_size)
             k = vvvv_p.shape[0]
             vvvv_p = vvvv_p.reshape(-1,nvir*nvir)
-            t[a:a+k] = np.dot(vvvv_p,t_amp_t).reshape(-1,nvir,nocc*nocc)
+            t[a:a+k] = np.dot(vvvv_p,t_amp).reshape(-1,nvir,nocc*nocc)
             del (vvvv_p)
```
Comment (susilehtola) [inline pyscf/adc/radc.py:477]:
```suggestion
            del vvvv_p
```

File: pyscf/adc/radc.py [hunk 11]
```diff
@@ -442,6 +518,37 @@ def density_matrix(myadc, T=None):
     return dm
 
 
+def analyze(myadc):
+
+    logger.info(myadc, "\n*************************************************************")
+    logger.info(myadc, "                Eigenvector analysis summary")
+    logger.info(myadc, "*************************************************************")
```
Comment (susilehtola) [inline pyscf/adc/radc.py:525]: The parentheses are a bit confusing. You might consider making this a multi-line string

```
logger.info(myadc, '''
*************************************************************
                Eigenvector analysis summary
*************************************************************''')
```

File: pyscf/adc/radc.py [hunk 12]
```diff
@@ -442,6 +518,37 @@ def density_matrix(myadc, T=None):
...
+
+    myadc.analyze_eigenvector()
+ 
+    if myadc.compute_properties:
+
+        logger.info(myadc, "\n*************************************************************")
+        logger.info(myadc, "            Spectroscopic factors analysis summary")
+        logger.info(myadc, "*************************************************************")
```
Comment (susilehtola) [inline pyscf/adc/radc.py:533]: Same here; would look cleaner with a multi-line string

File: pyscf/adc/radc.py [hunk 13]
```diff
@@ -1378,84 +1703,135 @@ def sigma_(r):
...
+               temp -=  0.25 * lib.einsum('mlab,jab->lmj',t2_1,r2)
+               temp +=  0.25 * lib.einsum('mlab,jba->lmj',t2_1,r2)
 
-               t2_1_a = t2_1 - t2_1.transpose(1,0,2,3).copy()
-               t2_1_a_t = t2_1_a.reshape(nocc,nocc,-1)
-               r2_a = r2_a.reshape(nocc,-1)
-               temp =  0.25 * lib.einsum('lmp,jp->lmj',t2_1_a_t,r2_a)
                s[s1:f1] += lib.einsum('lmj,lamj->a',temp, eris_ovoo, optimize=True)
                s[s1:f1] -= lib.einsum('lmj,malj->a',temp, eris_ovoo, optimize=True)
+               del (temp)
```
Comment (susilehtola) [inline pyscf/adc/radc.py:1735]:
```suggestion
               del temp
```

File: pyscf/adc/radc.py [hunk 14]
```diff
@@ -1378,84 +1703,135 @@ def sigma_(r):
...
+               temp_1 = -lib.einsum('lyd,lixd->ixy',temp,t2_1,optimize=True)
                s[s2:f2] -= temp_1.reshape(-1)
-               del eris_ovvv
+
+               del (temp_s_a)
+               del (temp_s_a_1)
+               del (temp_t2_r2_1)
+               del (temp_t2_r2_2)
+               del (temp_t2_r2_3)
+               del (temp_t2_r2_4)
```
Comment (susilehtola) [inline pyscf/adc/radc.py:1818]:
```suggestion
               del temp_s_a
               del temp_s_a_1
               del temp_t2_r2_1
               del temp_t2_r2_2
               del temp_t2_r2_3
               del temp_t2_r2_4
```

File: pyscf/adc/radc.py [hunk 15]
```diff
@@ -1378,84 +1703,135 @@ def sigma_(r):
...
+               temp  += lib.einsum('lxd,ilyd->ixy',temp_2_1,t2_1,optimize=True)
+               temp  -= lib.einsum('lxd,ildy->ixy',temp_2_1,t2_1,optimize=True)
                s[s2:f2] += temp.reshape(-1)
-
+              
+               del (t2_1)
+               del (temp)
+               del (temp_1)
+               del (temp_1_1)
+               del (temp_2_1)
```
Comment (susilehtola) [inline pyscf/adc/radc.py:1832]:
```suggestion
               del t2_1
               del temp
               del temp_1
               del temp_1_1
               del temp_2_1
```

File: pyscf/adc/uadc.py [hunk 16]
```diff
@@ -49,43 +49,47 @@ def kernel(adc, nroots=1, guess=None, eris=None, verbose=None):
...
-
-    spec_factors = adc.get_spec_factors(T, U, nroots)
+    if adc.compute_properties:
+        adc.P,adc.X = adc.get_properties(nroots)
 
     nfalse = np.shape(conv)[0] - np.sum(conv)
+
+    logger.info(adc, "\n*************************************************************")
+    logger.info(adc, "                  ADC calculation summary")
+    logger.info(adc, "*************************************************************")
```
Comment (susilehtola) [inline pyscf/adc/uadc.py:63]: Use multi-line string

File: pyscf/adc/uadc.py [hunk 17]
```diff
@@ -49,43 +49,47 @@ def kernel(adc, nroots=1, guess=None, eris=None, verbose=None):
...
-    if adc.verbose >= logger.INFO:
-        if nroots == 1:
-            logger.info(adc, '%s root %d    Energy (Eh) = %.10f    Energy (eV) = %.8f    Spec factors = %.8f    conv = %s',
-                         adc.method, 0, E, E*27.2114, spec_factors, conv)
-        else :
-            for n, en, pn, convn in zip(range(nroots), E, spec_factors, conv):
-                logger.info(adc, '%s root %d    Energy (Eh) = %.10f    Energy (eV) = %.8f    Spec factors = %.8f    conv = %s',
-                          adc.method, n, en, en*27.2114, pn, convn)
-        log.timer('ADC', *cput0)
+        logger.info(adc, "\n WARNING : Davidson iterations for " + str(nfalse) + " root(s) not converged\n")
```
Comment (susilehtola) [inline pyscf/adc/uadc.py:66]: logger.warn

File: pyscf/adc/uadc.py [hunk 18]
```diff
@@ -1829,89 +2111,91 @@ def ea_adc_diag(adc,M_ab=None,eris=None):
...
+#
+#            eris_oVoV_p = np.ascontiguousarray(eris_ooVV.transpose(0,2,1,3))
+#            eris_oVoV_p = eris_oVoV_p.reshape(nocc_a*nvir_b, nocc_a*nvir_b)
+#
+#            temp = np.zeros((nvir_a, nocc_a, nvir_b))
+#            temp[:] += np.diagonal(eris_oVoV_p).reshape(nocc_a,nvir_b)
+#            temp = np.ascontiguousarray(temp.transpose(1,2,0))
+#            diag[s_aba:f_aba] += -temp.reshape(-1)
+#        else :
+#           raise Exception("Precond not available for out-of-core and density-fitted algo")
```
Comment (susilehtola) [inline pyscf/adc/uadc.py:2199]: Why is this block commented?

File: pyscf/adc/uadc.py [hunk 19]
```diff
@@ -2006,100 +2290,103 @@ def ip_adc_diag(adc,M_ij=None,eris=None):
...
+#
+#            eris_OvOv_p = np.ascontiguousarray(eris_OOvv.transpose(0,2,1,3))
+#            eris_OvOv_p = eris_OvOv_p.reshape(nocc_b*nvir_a, nocc_b*nvir_a)
+#
+#            temp = np.zeros((nocc_a, nocc_b, nvir_a))
+#            temp[:] += np.diagonal(eris_OvOv_p).reshape(nocc_b,nvir_a)
+#            temp = np.ascontiguousarray(temp.transpose(2,1,0))
+#            diag[s_aba:f_aba] += temp.reshape(-1)
+#        else :
+#           raise Exception("Precond not available for out-of-core and density-fitted algo")
```
Comment (susilehtola) [inline pyscf/adc/uadc.py:2389]: Same here

File: pyscf/adc/uadc.py [hunk 20]
```diff
@@ -2285,31 +2566,91 @@ def sigma_(r):
...
+            s[s_a:f_a] += 0.5*lib.einsum('icab,ibc->a',eris_ovvv, r_aaa_[a:a+k], optimize = True)
+            s[s_a:f_a] -= 0.5*lib.einsum('ibac,ibc->a',eris_ovvv, r_aaa_[a:a+k], optimize = True)
+            temp[a:a+k] += lib.einsum('icab,a->ibc', eris_ovvv, r_a, optimize = True)
+            temp[a:a+k] -= lib.einsum('ibac,a->ibc', eris_ovvv, r_a, optimize = True)
+            del eris_ovvv
+            a += k
+
         s[s_aaa:f_aaa] += temp[:,ab_ind_a[0],ab_ind_a[1]].reshape(-1)
-        del eris_ovvv
+        del (temp)
```
Comment (susilehtola) [inline pyscf/adc/uadc.py:2591]:
```suggestion
        del temp
```

File: pyscf/adc/dfadc.py (hunk 1 above)
Comment (Samragni-93) [reply pyscf/adc/dfadc.py:46]: This function is used in the unrestricted code to calculate ovvv integrals with different spin cases. So, norb_1 can be nocc_alpha/ nocc_beta. Same for the virtual orbitals as well. This is why I have used generic variable names.

File: pyscf/adc/__init__.py (hunk 2 above)
Comment (Samragni-93) [reply pyscf/adc/__init__.py:44]: Had to comment out the ROHF functionality due to some erroneous results. Need to investigate further with convert_to_uhf , which we plan to do later on.

File: pyscf/adc/dfadc.py (hunk 3 above)
Comment (Samragni-93) [reply pyscf/adc/dfadc.py:18]: Added

File: pyscf/adc/dfadc.py (hunk 4 above)
Comment (Samragni-93) [reply pyscf/adc/dfadc.py:42]: `get_ovvv_df` is used in the restricted code, whereas `get_ovvv_a_df` is used in the unrestricted code to compute different spin cases of ovvv. I have added docstrings and renamed `get_ovvv_a_df` to `get_ovvv_spin_df`.

File: pyscf/adc/radc.py (hunk 7 above)
Comment (Samragni-93) [reply pyscf/adc/radc.py:108]: Yes, this is a conversion from chemists' to physicists' notation. Also, sliced transpose does not return a copy.

File: pyscf/adc/uadc.py (hunk 18 above)
Comment (Samragni-93) [reply pyscf/adc/uadc.py:2202]: This block includes the ADC(2)-x and ADC(3) terms into the preconditioner but based on our tests, that does not improve convergence, and sometimes can make things a little worse. So, we commented out the terms as of now.

**Comment by Samragni-93** [comment]
Thanks for the suggestions @susilehtola. I have addressed most of the changes via [3c34482](https://github.com/pyscf/pyscf/pull/795/commits/3c34482baf091895bce07c2877f0781e06bbbdf1). The PR is ready for merge now.

File: pyscf/adc/dfadc.py (hunk 1 above)
Comment (susilehtola) [reply pyscf/adc/dfadc.py:46]: If `norb_1` is always `nocc_alpha` or `nocc_beta`, why not just name it `nocc`? Same for the virtual orbitals.

Comment (Samragni-93) [reply pyscf/adc/dfadc.py:46]: Yeah, norb_1 can be named nocc. But norb_2 and norb_3 will be dependent on Lov and Lvv, respectively. In that case, norb_2 can be named nvir_1, and norb_3 can be named nvir_2. I can make these changes.

**Comment by ssh2** [comment]
Found a bug in the UADC eigenvector analysis for open-shell molecules. Fixed via 1d2ac2eb4308f4bdf1e9d9edf3f15c636543af89. I think this PR should be finally ready for merge.

**Comment by sokolov-group** [comment]
Commit 2b24873a7bd6e379ce51704647252929dff09e11 was accidentally added to this PR. Commit 2c4b6cfce72b092c548041db59b808fb063142fc reverts that change. Since this PR has been open for a while and no other comments have been made, I am merging this PR to dev.

**Pull request merged**